README.md
# SIT-PORTAIL

Portail des services TI de l'Université Laval

## Prérequis pour le développement

- Docker (Windows ou Linux)
- WSL 2 avec ubuntu 20.04 LTS (https://docs.microsoft.com/en-us/windows/wsl/tutorials/wsl-containers)
- VSCode

## Exécuter et développer localement

Faire le git clone dans votre home de ubuntu dans WSL2

```bash
cd sit-gabarit
code .
```

Dans VSCode

- Ctrl+Shift+P
- Remote-Containers: Open Folder in Container...
- Dans le terminal exécuter "npm start"
- Lancez "Launch Chrome" dans les options d'exécutions pour debugger.

## Conventions

### Nom de branches

Une branche devrait être nommé ainsi: _type_/_descriptif_, où _descriptif_ est un court titre décrivant le problème adressé par la branche. Si une demande ou une épopée est reliée, la JIRA peut être spécifiée.

Types disponibles:

- feature: Pour toute nouvelle fonctionnalité ajoutée
- bugfix: Pour appliquer un changement suite à une MR rejeté, une démo ou un test en dev/approbation
- hotfix: Pour tout changement rapide ou critique suite à un incident
- build: Pour des changements reliés au build ou au déploiement, sans impact sur l'application.
- Rares:
  - misc: Pour tester une fonctionnalité ou un nouvel outil. Une branche pour s'amuser et expérimenter.
  - merge: Pour résoudre les merge conflitcs

## Déploiement de l'application

1. Appliquer des modifications au code, au Dockerfile ou aux configurations de node.js (package.json)
2. Pousser le code sur le dépôt dans la branche `dev`
3. **Étape automatique**: Déclencher le pipeline sur [le projet Jenkins](https://jenkins.svc.ulaval.ca/job/fondation/job/SIT/job/sit-portail/job/SIT-Portail-dv/)
   > Il est possible d'automatiser le déclenchement du pipeline lors du push sur le dépôt. Voir la procédure [GEG-PE Déclencher un build Jenkins depuis Gitlab (DCO)](https://wiki.dti.ulaval.ca/pages/viewpage.action?pageId=123476253)

### Description des étapes du pipeline de déploiement

1. Build de l'application sur nodejs 14 avec `npm ci` (installe les dépendances) et `npm run build`
2. Build de l'image Docker nginx qui lancera l'application
3. Push de l'image sur le registre d'Artifactory (maven.ulaval.ca)
4. Récupération de l'identifiant (digest) de l'image sur artifactory. Cet identifiant est utilisé par openshift pour qu'on référence exactement l'image qu'on veut
5. Traitement des tags. Cette étape sert à classer les tags.
   > On garde un historique de trois version dans artifactory accessible avec des tags (dv, dv-1 et dv-2). Par contre, pour faire un retour en arrière dans openshift, on n'as qu'à faire un `oc rollback`
6. Déploiement des configurations Openshift: Tout est traité par code à l'aide du fichier sit-portail.yml. Les différents paramètres sont gérés dans le Jenkinsfile
7. Déploiement de l'application: On fait un oc rollout pour déployer la nouvelle configuration

## Scripts disponibles

Dans le dossier du projet, vous pouvez lancer les scripts suivants:

### `npm start`

Lance l'application en mode développement afin de la tester.<br>
Ouvrir [http://localhost:8080](http://localhost:8080) pour la consulter dans le navigateur

> Le port 8080 a été configuré dans `package.json`

The page will reload if you make edits.<br>
You will also see any lint errors in the console.

### `npm run start-pc`

Même chose que `npm start`, mais la commande est modifiée pour fonctionner avec dos et Powershell<br>

### `npm test`

Lance les tests unitaires (présents dans les fichiers `.test.js`).<br>
Pour plus d'informations sur la création des tests unitaires en react, voir la page du wiki [Tests en REACT](https://wiki.dti.ulaval.ca/display/SIT/Tests+en+REACT>) <br>
See the section about [running tests](https://facebook.github.io/create-react-app/docs/running-tests) for more information.

### `npm run build`

Créer le build de production de l'application

Builds the app for production to the `build` folder.<br>
It correctly bundles React in production mode and optimizes the build for the best performance.

The build is minified and the filenames include the hashes.<br>
Your app is ready to be deployed!

See the section about [deployment](https://facebook.github.io/create-react-app/docs/deployment) for more information.

### `npm run eject`

**Note: this is a one-way operation. Once you `eject`, you can’t go back!**

If you aren’t satisfied with the build tool and configuration choices, you can `eject` at any time. This command will remove the single build dependency from your project.

Instead, it will copy all the configuration files and the transitive dependencies (Webpack, Babel, ESLint, etc) right into your project so you have full control over them. All of the commands except `eject` will still work, but they will point to the copied scripts so you can tweak them. At this point you’re on your own.

You don’t have to ever use `eject`. The curated feature set is suitable for small and middle deployments, and you shouldn’t feel obligated to use this feature. However we understand that this tool wouldn’t be useful if you couldn’t customize it when you are ready for it.


src/CHANGELOG.md
<!-- markdownlint-disable MD024 -->

# Historique des modifications aux services TI de l'Université Laval

Le format est basé sur [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
et ce projet adhère au [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.26.0] - 2021-11-03

### Ajout

#### Serveurs

- Prise en charge des nouvelles plages de mise à jour à la création de serveurs.
- Ajout d'un filtre sur le type de serveur dans l'inventaire.

### Modification

#### Serveurs

- Les serveurs vmware sont affichés par défaut dans l'inventaire des serveurs.

## [1.25.0] - 2021-10-25

### Ajout

#### Portail

- Ajout d'une page 404.

### Modification

#### Serveurs

- Ajout d'une notification de succès pour la suppression de serveur.

### Correctif

#### Portail

- Correction du bogue avec l'expiration du token.
- Correction du bogue sur l'ordre de disparition des notifications.

#### Serveurs

- Correction du bogue pour la création multiple de serveurs.
- Correction du bogue sur les actions applicables à plusieurs serveurs.

## [1.24.0] - 2021-09-22

### Modification

#### Serveurs

- Gain de performance obtenu grâce au préchargement des VLANs dans l’assistant de création de serveur.

### Correctif

#### Portail

- Correction du bogue avec le jeton d'authentification au démarrage du Portail.

#### Postes et applications virtuelles

- Correction du bogue d'affichage pour le Idle Time des connections.

## [1.23.0] - 2021-09-15

### Modification

#### Portail

- Standardisation des pages d'inventaire.
- Standardisation des icônes de création.
- Standardisation des icônes d'inventaire.

### Correctif

#### Portail

- Correction d'erreurs d'orthographe.

## [1.22.0] - 2021-08-24

### Ajout

#### Portail

- Nouvelle section pour les notifications.
- Nouvelle URL servicesti.ulaval.ca pour accéder au portail.

### Modification

#### Portail

- Nouvelle page d'accueil.
- Nouvelle page de commentaires.

### Suppression

#### Serveurs

- Retrait de l'onboarding Azure.

### Correctif

#### Serveurs

- Validation de l'espace disque et du nombre de CPU.
- Correction du bogue d'affichage des disques.

## [1.21.0] - 2021-07-15

### Modification

#### Serveurs

- Affichage de la sélection de la plage des mises à jour ouverte par défaut.
- Refonte du tableau de sélection des VLANs.

### Correctif

#### Postes et applications virtuelles

- Correction des droits d'accès pour PTV.

## [1.20.0] - 2021-07-07

### Ajout

#### Serveurs

- Fonctionnalité de conformité (SCCM) pour les serveurs Windows.

#### Postes et applications virtuelles

- Permettre la modification des permissions dans les remoteApps.

### Correctif

#### Serveurs

- Correction d'un bogue d'affichage pour les OS.

## [1.19.0] - 2021-06-07

### Ajout

#### Portail

- Ajout du support mobile pour l'entête du site.

### Modification

#### Portail

- Amélioration du menu de l'utilisateur.

## [1.18.0] - 2021-06-02

### Ajout

#### Portail

- Ajoute une barre de déroulement dans la section versions, dans l'inventaire des serveurs et l'inventaire PTV.

### Modification

#### Serveurs

- Validation du nom du serveur à créer après un certains temps d'inactivité à l'étape d'identification. Il n'est plus nécessaire de devoir changer le focus pour déclencher la validation.

### Suppression

#### Portail

- Retrait de la fonctionnalité "Vider la cache".

- Retrait de la fonctionnalité "Table des paramètres".

## [1.17.0] - 2021-05-24

### Ajout

#### Portail

- Ajout de l'historique des modifications au service d'infrastructure technologique.

#### Serveurs

- Ajout du positionnement du snapshot.

### Modification

#### Serveurs

- Désactivation des champs CPU et RAM en mode modification. Ils seront réactivés lorsque la fonctionnalité sera disponible.

## [1.16.0] - 2021-05-14

### Ajout

#### Portail

- Ajout de la nouvelle section "Aide" dans le menu du haut qui contient un message concernant l'accessibilité Web.
- Cliquer sur le logo de l'Université Laval en haut à gauche redirige maintenant vers la page d'accueil.

#### Serveurs

- Ajout de la fonctionnalité de retour arrière sur les snaphosts de VM.

## [1.15.0] - 2021-04-30

### Ajout

#### Serveurs

- Ajout de la validation de la mémoire vive des VMs selon l'OS sélectionné.

#### Postes et applications virtuelles

- Ajout de la création de collections.
- Ajout de l'affichage des connexions client.
- Ajout de la recherche de serveurs.

### Correctif

#### Portail

- Corrige la couleur du texte des boutons primaires.

#### Bases de données

- Correction du CSV produit.

## [1.14.0] - 2021-04-16

### Ajout

#### Serveurs

- Ajout de la mise à jour des plages de maintenance des serveurs Windows existants.

### Correctif

#### Portail

- Les textes du menu de gauche (hamburger) sont désormais tous alignés à gauche.

#### Serveurs

- Correction d'une VM sans OS.

## [1.13.0] - 2021-04-02

### Ajout

#### Postes et applications virtuelles

- Ajout de la validation des utilisateurs dans PTV quand on les ajoute à une collection.

#### Bases de données

- Ajout de la sortie CSV et Jira.

### Modification

#### Serveurs

- Change le label pour les machines sans domaines (AUCUN à EXTERNE).

### Correctif

#### Portail

- Correction de grammaire.

#### Serveurs

- Correction initialisation serveur physique.

## [1.12.0] - 2021-03-19

### Ajout

#### Bases de données

- Fonctionnalité de gestion des utilisateurs BD.

### Modification

#### Postes et applications virtuelles

- Le bouton final de l'assistant redirige vers la nouvelle collection.

### Suppression

#### Serveurs

- Retrait de l'option McAfee.

### Correctif

#### Serveurs

- Correction du message de la toast d'erreur lors de la réinstallation de serveur.

#### Bases de données

- Correction de grammaire.

## [1.11.0] - 2021-03-05

### Ajout

#### Postes et applications virtuelles

- Validation du format des utilisateurs.

### Correctif

#### Serveurs

- Correction du problème de la réinstallation de serveurs physique.

src/service-worker.ts
/// <reference lib="webworker" />
/* eslint-disable no-restricted-globals */

// This service worker can be customized!
// See https://developers.google.com/web/tools/workbox/modules
// for the list of available Workbox modules, or add any other
// code you'd like.
// You can also remove this file if you'd prefer not to use a
// service worker, and the Workbox build step will be skipped.

import { clientsClaim } from 'workbox-core';
import { ExpirationPlugin } from 'workbox-expiration';
import { precacheAndRoute, createHandlerBoundToURL } from 'workbox-precaching';
import { registerRoute } from 'workbox-routing';
import { StaleWhileRevalidate } from 'workbox-strategies';

declare const self: ServiceWorkerGlobalScope;

clientsClaim();

// Precache all of the assets generated by your build process.
// Their URLs are injected into the manifest variable below.
// This variable must be present somewhere in your service worker file,
// even if you decide not to use precaching. See https://cra.link/PWA
// eslint-disable-next-line no-underscore-dangle
precacheAndRoute(self.__WB_MANIFEST);

// Set up App Shell-style routing, so that all navigation requests
// are fulfilled with your index.html shell. Learn more at
// https://developers.google.com/web/fundamentals/architecture/app-shell
const fileExtensionRegexp = new RegExp('/[^/?]+\\.[^/]+$');
registerRoute(
  // Return false to exempt requests from being fulfilled by index.html.
  ({ request, url }: { request: Request; url: URL }) => {
    // If this isn't a navigation, skip.
    if (request.mode !== 'navigate') {
      return false;
    }

    // If this is a URL that starts with /_, skip.
    if (url.pathname.startsWith('/_')) {
      return false;
    }

    // If this looks like a URL for a resource, because it contains
    // a file extension, skip.
    if (url.pathname.match(fileExtensionRegexp)) {
      return false;
    }

    // Return true to signal that we want to use the handler.
    return true;
  },
  createHandlerBoundToURL(`${process.env.PUBLIC_URL}'/index.html'`)
);

// An example runtime caching route for requests that aren't handled by the
// precache, in this case same-origin .png requests like those from in public/
registerRoute(
  // Add in any other file extensions or routing criteria as needed.
  ({ url }) =>
    url.origin === self.location.origin && url.pathname.endsWith('.png'),
  // Customize this strategy as needed, e.g., by changing to CacheFirst.
  new StaleWhileRevalidate({
    cacheName: 'images',
    plugins: [
      // Ensure that once this runtime cache reaches a maximum size the
      // least-recently used images are removed.
      new ExpirationPlugin({ maxEntries: 50 }),
    ],
  })
);

// This allows the web app to trigger skipWaiting via
// registration.waiting.postMessage({type: 'SKIP_WAITING'})
self.addEventListener('message', (event) => {
  if (event.data && event.data.type === 'SKIP_WAITING') {
    self.skipWaiting();
  }
});

// Any other custom service worker logic can go here.


src/serviceWorkerRegistration.ts
// This optional code is used to register a service worker.
// register() is not called by default.

// This lets the app load faster on subsequent visits in production, and gives
// it offline capabilities. However, it also means that developers (and users)
// will only see deployed updates on subsequent visits to a page, after all the
// existing tabs open on the page have been closed, since previously cached
// resources are updated in the background.

// To learn more about the benefits of this model and instructions on how to
// opt-in, read https://cra.link/PWA

const isLocalhost = Boolean(
  window.location.hostname === 'localhost' ||
    // [::1] is the IPv6 localhost address.
    window.location.hostname === '[::1]' ||
    // 127.0.0.0/8 are considered localhost for IPv4.
    window.location.hostname.match(
      /^127(?:\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}$/
    )
);

type Config = {
  onSuccess?: (registration: ServiceWorkerRegistration) => void;
  onUpdate?: (registration: ServiceWorkerRegistration) => void;
};

export function unregister(): void {
  if ('serviceWorker' in navigator) {
    navigator.serviceWorker.ready.then((registration) => {
      registration.unregister();
    });
  }
}

function registerValidSW(swUrl: string, config?: Config) {
  navigator.serviceWorker.register(swUrl).then((registration) => {
    // eslint-disable-next-line no-param-reassign
    registration.onupdatefound = () => {
      const installingWorker = registration.installing;
      if (installingWorker == null) {
        return;
      }
      installingWorker.onstatechange = () => {
        if (installingWorker.state === 'installed') {
          if (navigator.serviceWorker.controller) {
            // At this point, the updated precached content has been fetched,
            // but the previous service worker will still serve the older
            // content until all client tabs are closed.

            // Execute callback
            if (config && config.onUpdate) {
              config.onUpdate(registration);
            }
          } else if (config && config.onSuccess) {
            config.onSuccess(registration);
          }
        }
      };
    };
  });
}

function checkValidServiceWorker(swUrl: string, config?: Config) {
  // Check if the service worker can be found. If it can't reload the page.
  fetch(swUrl, {
    headers: { 'Service-Worker': 'script' },
  }).then((response) => {
    // Ensure service worker exists, and that we really are getting a JS file.
    const contentType = response.headers.get('content-type');
    if (
      response.status === 404 ||
      (contentType != null && contentType.indexOf('javascript') === -1)
    ) {
      // No service worker found. Probably a different app. Reload the page.
      navigator.serviceWorker.ready.then((registration) => {
        registration.unregister().then(() => {
          window.location.reload();
        });
      });
    } else {
      // Service worker found. Proceed as normal.
      registerValidSW(swUrl, config);
    }
  });
}

export function register(config?: Config): void {
  if (process.env.NODE_ENV === 'production' && 'serviceWorker' in navigator) {
    // The URL constructor is available in all browsers that support SW.
    const publicUrl = new URL(process.env.PUBLIC_URL, window.location.href);
    if (publicUrl.origin !== window.location.origin) {
      // Our service worker won't work if PUBLIC_URL is on a different origin
      // from what our page is served on. This might happen if a CDN is used to
      // serve assets; see https://github.com/facebook/create-react-app/issues/2374
      return;
    }

    window.addEventListener('load', () => {
      const swUrl = `${process.env.PUBLIC_URL}/service-worker.js`;

      if (isLocalhost) {
        // This is running on localhost. Let's check if a service worker still exists or not.
        checkValidServiceWorker(swUrl, config);

        // Add some additional logging to localhost, pointing developers to the
        // service worker/PWA documentation.
        navigator.serviceWorker.ready.then(() => {});
      } else {
        // Is not localhost. Just register service worker
        registerValidSW(swUrl, config);
      }
    });
  }
}
